src/libexpr/primops/fromYAML.cc

```diff
@@ -0,0 +1,156 @@
+#ifdef HAVE_RYML
+
+#include "primops.hh"
+#include "eval-inline.hh"
+
+#include <ryml.hpp>
+#include <c4/format.hpp>
+#include <c4/std/string.hpp>
+
+
+namespace nix {
+
+struct NixContext {
+    EvalState & state;
+    const PosIdx pos;
+    std::string_view yaml;
+};
+
+static void s_error [[ noreturn ]] (const char* msg, size_t len, ryml::Location, void *nixContext)
+{
+    auto context = static_cast<const NixContext *>(nixContext);
+    if (nixContext) {
+        throw EvalError({
+            .msg = hintfmt("while parsing the YAML string '%1%':\n\n%2%",
+                context->yaml, std::string_view(msg, len)),
+            .errPos = context->state.positions[context->pos]
+        });
+    } else {
+        throw EvalError({
+            .msg = hintfmt("failed assertion in rapidyaml library:\n\n%1%",
+                std::string_view(msg, len))
+        });
+    }
+}
+
+static void visitYAMLNode(NixContext & context, Value & v, ryml::ConstNodeRef t) {
+
+    auto valTypeCheck = [=] (ryml::YamlTag_e tag, bool defaultVal = true) {
+        auto valTag = ryml::TAG_NONE;
+        if (t.has_val_tag()) {
+            auto tag = t.val_tag();
+            valTag = tag == "!" ? ryml::TAG_STR : ryml::to_tag(tag);
+        }
+        return valTag == ryml::TAG_NONE ? defaultVal : valTag == tag;
+    };
+
+    v.mkBlackhole();
```

Comment:
```suggestion
```
We shouldn't rely on `mkBlackhole()` leaving the value as a thunk. It's better to handle the fall-through case (i.e when the YAML cannot be represented) in another way (maybe by simplifying the logic into a single "if ... else" chain).